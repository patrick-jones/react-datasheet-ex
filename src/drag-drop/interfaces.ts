import RDS from 'react-datasheet';
import {ConnectDragPreview, ConnectDragSource, ConnectDropTarget} from 'react-dnd';
import {ColumnHeader, HeaderRendererProps, RowRendererProps} from '../interfaces';

/**
 * Properties that must be provided to a drag and drop enabled row
 */
export interface DragDropRowProps<T extends RDS.Cell<T, V>, V = string> {
  /** The current row index */
  row: number;
  /** Callback for when a row is dropped on another */
  readonly onRowDropped: (source: number, target: number) => void;
}

/**
 * Properties that must be provided to a drag and drop enabled column header
 */
export interface DragDropHeaderProps {
  /** The current column index */
  col: number;
  /** Callback for one header is dropped on another */
  readonly onHeaderDropped: (source: number, target: number) => void;
}

/**
 * Drag source properties that are injected into a drag and drop provider
 */
export interface InjectedDragSourceProps {
  readonly connectDragSource: ConnectDragSource;
  readonly connectDragPreview: ConnectDragPreview;
}

/**
 * Drop target properties that are injected into a drag and drop provider
 */
export interface InjectedDropTargetProps {
  readonly connectDropTarget: ConnectDropTarget;
  readonly isDropOver: boolean;
}

/**
 * Combined drag source and drop target properties that are injected
 * into a drag and drop provider
 */
export interface InjectedDragDropProps extends InjectedDragSourceProps, InjectedDropTargetProps {}

/**
 * Extends `ReactDatasheet.RowRendererProps` to include
 * drag and drop properties
 */
export interface DragDropRowRendererProps<CellType, V = string>
  extends RowRendererProps<CellType, V>,
    DragDropRowProps<CellType, V> {}

/**
 * Extends `HeaderRendererProps` to include drag and drop properties.
 */
export interface DragDropHeaderRendererProps<T extends ColumnHeader = ColumnHeader>
  extends HeaderRendererProps<T>,
    DragDropHeaderProps {}
