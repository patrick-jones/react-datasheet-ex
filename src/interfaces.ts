import React, {CSSProperties} from 'react';
import RDS from 'react-datasheet';

export interface RowLocation {
  row: number;
}

export interface ColumnLocation {
  col: number;
}

export interface ColumnHeader {
  id: string;
  title: string;
  width?: number;
}

export interface HeaderRendererProps<T extends ColumnHeader = ColumnHeader> extends ColumnLocation {
  header: T;
}

export type HeaderRenderer<T extends ColumnHeader = ColumnHeader,
  R extends HeaderRendererProps<T> = HeaderRendererProps<T>> = React.ComponentType<R>;

export type HeaderContentRenderer<T extends ColumnHeader = ColumnHeader>
  = React.ComponentType<HeaderRendererProps<T>>;

export interface SheetRendererProps<
    T extends RDS.Cell<T, V>, V = string, H extends ColumnHeader = ColumnHeader
  >
  extends RDS.SheetRendererProps<T, V> {
  /** React style object that should be applied to the enclosing table */
  style?: CSSProperties | undefined;
  /**
   * An array of `ColumnHeader`, which are rendered into `th` elements
   * @default []
   */
  headers: H[];
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderRenderer
   */
  headerRenderer: HeaderRenderer;
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderContentRenderer
   */
  headerContentRenderer: HeaderContentRenderer;
}


export type ActionContentRenderer<T extends RowLocation = RowLocation>
  = React.ComponentType<T>;

export interface RowRendererProps<
    T extends RDS.Cell<T, V>, V = string, AP extends RowLocation = RowLocation
  > extends RDS.RowRendererProps<T, V> {

  actionContentRenderer?: ActionContentRenderer<AP>;
}
