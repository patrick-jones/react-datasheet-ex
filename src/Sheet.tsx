import React, {PureComponent, ReactNode} from 'react';
import RDS from 'react-datasheet';

import {
  ColumnHeader,
  HeaderRenderer,
  SheetRendererProps,
} from './interfaces';
import HeaderCell from './HeaderCell';
import HeaderTitle from './HeaderTitle';

import {READ_ONLY_CELL, ACTION_CELL} from './ClassNames';

// import './react-datasheet-ex.css';


const actionClassNames = `${READ_ONLY_CELL} ${ACTION_CELL}`;

/**
 * A custom `sheetRenderer` for `react-datasheet` that renders a sheet
 * with column headers and an leading 'action" cell on each row.
 */
export default class Sheet<
    T extends RDS.Cell<T, V>, V = string, H extends ColumnHeader = ColumnHeader
  > extends PureComponent<SheetRendererProps<T, V, H>> {

  static readonly defaultProps = {
    headers: [],
    headerCellRenderer: HeaderCell,
    headerRenderer: HeaderTitle,
    overflow: 'clip',
  };

  render(): ReactNode {
    const {
      className,
      headers,
      headerCellRenderer: HeaderCellRenderer,
      headerRenderer: HeaderRenderer,
      overflow,
      style,
    } = this.props;

    return (
      <table className={className} style={style}>
        <thead>
          <tr key='$headers'>
            <th className={actionClassNames} key='$$actionCell'/>
            {headers.map((header: H, col: number) => {
              const hrProps = {col, header, overflow};
              return (
                <HeaderCellRenderer
                  key={`header-${header.id}`}
                  {...hrProps}
                >
                  <HeaderRenderer key={header.id} {...hrProps} />
                </HeaderCellRenderer>
              );
            })}
          </tr>
        </thead>
        <tbody>
        {this.props.children}
        </tbody>
      </table>
    );
  }
}
