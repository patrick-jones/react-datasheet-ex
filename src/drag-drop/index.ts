import DragDropHeaderProvider from './DragDropHeaderProvider';
import DragDropHeaderRenderer from './DragDropHeaderRenderer';
import DragDropRowProvider from './DragDropRowProvider';
import DragDropRowRenderer from './DragDropRowRenderer';

export {DragDropHeaderProvider, DragDropHeaderRenderer, DragDropRowProvider, DragDropRowRenderer};
