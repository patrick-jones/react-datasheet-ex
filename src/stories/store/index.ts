import {model} from './model';
import {ExampleModel} from './interfaces';
import {Actions, SheetActionCreators} from './actions';
import reducer from './reducer';

export {model};

export function preprocess(model: ExampleModel, ...actions: Actions<SheetActionCreators>[]) {
  return actions.reduce(
    (reply, action) => {
      const nextState = reducer(reply, action);
      return Object.assign({}, reply, nextState);
    },
    model
  );
}
