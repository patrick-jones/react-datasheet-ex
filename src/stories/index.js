import React from 'react';
import {storiesOf} from '@storybook/react';
import {withState} from '@dump247/storybook-state';
import {doc} from 'storybook-readme';
import {withDocs} from './withDocs';

import {model, preprocess} from './store';

import '!style-loader!css-loader!react-datasheet/lib/react-datasheet.css';
import '../react-datasheet-ex.css';
import './stories.css';

import Readme from '../../README.md';
import SheetDocs from './Sheet.md';
import SheetStory from './Sheet.story';
import RowDocs from './Row.md';
import RowStory from './Row.story';
import HeaderDocs from './HeaderComponents.md';
import HeaderStory from './HeaderComponents.story';

import DragDropProvidersDocs from './DragDropProviders.md';
import DragDropProvidersStory from './DragDropProviders.story';
import DragDropRenderersDocs from './DragDropRenderers.md';
import DragDropRenderersStory from './DragDropRenderers.story';

import ResizableDocs from './Resizable.md';
import ResizableHeaderDocs from './ResizableHeader.md';
import ResizableHeaderStory from './ResizableHeader.story';
import ResizableHeaderContainerDocs from './ResizableHeaderContainer.md';
import ResizableHeaderContainerStory from './ResizableHeaderContainer.story';
import WidthResizeHandleDocs from './WidthResizeHandle.md';
import WidthResizeHandleStory from './WidthResizeHandle.story';
import {SheetActionCreators} from './store/actions';

const stories = storiesOf('ReactDatasheet Extensions', module);
const basic = storiesOf('Basic Components', module);
const dragDrop = storiesOf('Drag and Drop', module);
const resize = storiesOf('Resizable', module);

stories
  .add('Introduction', doc(Readme))
;

basic
  .add('Sheet',
    withDocs(SheetDocs,
      withState(model)(SheetStory)))
  .add('Row',
    withDocs(RowDocs,
      withState(model)(RowStory)))
  .add('Header Components',
    withDocs(HeaderDocs,
      withState(model)(HeaderStory)))
;

dragDrop
  .add('Drag and Drop Renderers',
    withDocs(DragDropRenderersDocs,
      withState(model)((DragDropRenderersStory))))
  .add('Drag and Drop Providers',
    withDocs(DragDropProvidersDocs,
      withState(model)((DragDropProvidersStory))))
;

const actions = model.headers.map((h, index) => SheetActionCreators.headerResizeEnd(index, 100));
const resizeModel = preprocess(model, ...actions);

resize
  .add('Overview', doc(ResizableDocs))
  .add('Resizable Header',
    withDocs(ResizableHeaderDocs,
      withState(resizeModel)((ResizableHeaderStory))))
  .add('Resizable Container',
    withDocs(ResizableHeaderContainerDocs,
      withState(resizeModel)((ResizableHeaderContainerStory))))
  .add('Width Resize Handle',
    withDocs(WidthResizeHandleDocs,
      withState(resizeModel)((WidthResizeHandleStory))))
;
