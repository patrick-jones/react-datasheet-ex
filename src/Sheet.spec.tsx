import React from 'react';
import {shallow} from 'enzyme';

import Sheet from './Sheet';
import {DragDropHeader} from './drag-drop';
import {model} from './stories/store';
import {ExampleCellType} from './stories/store/interfaces';


describe('Sheet basic rendering', () => {
  const {rows, headers} = model;
  const data: ExampleCellType[][] = rows.map(r => r.data);

  it('Should render', () => {
    const wrapper = shallow(
      <Sheet
        data={data}
        className='foobar'
        headers={headers}
      >
        <tr>
          <td colSpan={4}>shallow</td>
        </tr>
      </Sheet>
    );
    // console.log(wrapper.debug());
    expect(wrapper.find('thead tr').length).toBe(1);
    expect(wrapper.find('th').length).toBe(1);
    expect(wrapper.find('HeaderCell').length).toBe(4);
    expect(wrapper.find('HeaderTitle').length).toBe(4);
    expect(wrapper.find('tbody tr td').text()).toBe('shallow');
  });

  it('Matches the snapshot', () => {
    const wrapper = shallow(
      <Sheet
        data={data}
        className='foobar'
        headers={headers}
      >
        <tr>
          <td colSpan={4}>shallow</td>
        </tr>
      </Sheet>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('Correctly renders optional properties', () => {
    const wrapper = shallow(
      <Sheet
        data={data}
        className='foobar'
        style={{width: '100%'}}
        headers={headers}
        headerRenderer={hrProps => (
          <DragDropHeader
            {...hrProps}
            onHeaderDropped={(...args) => {}}
          >
            {hrProps.header.id}
          </DragDropHeader>
        )}
        headerCellRenderer={hrProps => (
          <th>{hrProps.children}</th>
        )}
      >
        <tr>
          <td colSpan={4}>shallow</td>
        </tr>
      </Sheet>
    );
    // console.log(wrapper.debug());
    expect(wrapper.find('thead tr').length).toBe(1);
    expect(wrapper.find('th').length).toBe(1);
    expect(wrapper.find('headerCellRenderer').length).toBe(4);
    expect(wrapper.find('headerRenderer').length).toBe(4);
    expect(wrapper.find('tbody tr td').text()).toBe('shallow');
    expect(wrapper.prop('style')).toEqual({width: '100%'});
  });

});
