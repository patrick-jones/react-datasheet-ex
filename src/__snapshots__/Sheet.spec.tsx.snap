// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sheet basic rendering Matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Sheet
    className="foobar"
    data={
      Array [
        Array [
          Object {
            "value": 1,
          },
          Object {
            "value": 2,
          },
          Object {
            "value": 3,
          },
          Object {
            "value": 4,
          },
        ],
        Array [
          Object {
            "value": 5,
          },
          Object {
            "value": 6,
          },
          Object {
            "value": 7,
          },
          Object {
            "value": 8,
          },
        ],
        Array [
          Object {
            "value": 9,
          },
          Object {
            "value": 10,
          },
          Object {
            "value": 11,
          },
          Object {
            "value": 12,
          },
        ],
      ]
    }
    headerCellRenderer={[Function]}
    headerRenderer={[Function]}
    headers={
      Array [
        Object {
          "align": "left",
          "id": "A",
          "title": "Left",
          "width": 80,
        },
        Object {
          "align": "right",
          "id": "B",
          "title": "Right",
        },
        Object {
          "id": "C",
          "title": "Center",
        },
        Object {
          "id": "D",
          "title": "Long Title that Might Overflow",
        },
      ]
    }
    overflow="clip"
  >
    <tr>
      <td
        colSpan={4}
      >
        shallow
      </td>
    </tr>
  </Sheet>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <thead>
          <tr>
            <th
              className="cell read-only rdx-action-cell"
            />
            <HeaderCell
              col={0}
              header={
                Object {
                  "align": "left",
                  "id": "A",
                  "title": "Left",
                  "width": 80,
                }
              }
              overflow="clip"
            >
              <HeaderTitle
                className=""
                col={0}
                header={
                  Object {
                    "align": "left",
                    "id": "A",
                    "title": "Left",
                    "width": 80,
                  }
                }
                overflow="clip"
              />
            </HeaderCell>
            <HeaderCell
              col={1}
              header={
                Object {
                  "align": "right",
                  "id": "B",
                  "title": "Right",
                }
              }
              overflow="clip"
            >
              <HeaderTitle
                className=""
                col={1}
                header={
                  Object {
                    "align": "right",
                    "id": "B",
                    "title": "Right",
                  }
                }
                overflow="clip"
              />
            </HeaderCell>
            <HeaderCell
              col={2}
              header={
                Object {
                  "id": "C",
                  "title": "Center",
                }
              }
              overflow="clip"
            >
              <HeaderTitle
                className=""
                col={2}
                header={
                  Object {
                    "id": "C",
                    "title": "Center",
                  }
                }
                overflow="clip"
              />
            </HeaderCell>
            <HeaderCell
              col={3}
              header={
                Object {
                  "id": "D",
                  "title": "Long Title that Might Overflow",
                }
              }
              overflow="clip"
            >
              <HeaderTitle
                className=""
                col={3}
                header={
                  Object {
                    "id": "D",
                    "title": "Long Title that Might Overflow",
                  }
                }
                overflow="clip"
              />
            </HeaderCell>
          </tr>
        </thead>,
        <tbody>
          <tr>
            <td
              colSpan={4}
            >
              shallow
            </td>
          </tr>
        </tbody>,
      ],
      "className": "foobar",
      "style": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <tr>
            <th
              className="cell read-only rdx-action-cell"
            />
            <HeaderCell
              col={0}
              header={
                Object {
                  "align": "left",
                  "id": "A",
                  "title": "Left",
                  "width": 80,
                }
              }
              overflow="clip"
            >
              <HeaderTitle
                className=""
                col={0}
                header={
                  Object {
                    "align": "left",
                    "id": "A",
                    "title": "Left",
                    "width": 80,
                  }
                }
                overflow="clip"
              />
            </HeaderCell>
            <HeaderCell
              col={1}
              header={
                Object {
                  "align": "right",
                  "id": "B",
                  "title": "Right",
                }
              }
              overflow="clip"
            >
              <HeaderTitle
                className=""
                col={1}
                header={
                  Object {
                    "align": "right",
                    "id": "B",
                    "title": "Right",
                  }
                }
                overflow="clip"
              />
            </HeaderCell>
            <HeaderCell
              col={2}
              header={
                Object {
                  "id": "C",
                  "title": "Center",
                }
              }
              overflow="clip"
            >
              <HeaderTitle
                className=""
                col={2}
                header={
                  Object {
                    "id": "C",
                    "title": "Center",
                  }
                }
                overflow="clip"
              />
            </HeaderCell>
            <HeaderCell
              col={3}
              header={
                Object {
                  "id": "D",
                  "title": "Long Title that Might Overflow",
                }
              }
              overflow="clip"
            >
              <HeaderTitle
                className=""
                col={3}
                header={
                  Object {
                    "id": "D",
                    "title": "Long Title that Might Overflow",
                  }
                }
                overflow="clip"
              />
            </HeaderCell>
          </tr>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": "$headers",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <th
                className="cell read-only rdx-action-cell"
              />,
              Array [
                <HeaderCell
                  col={0}
                  header={
                    Object {
                      "align": "left",
                      "id": "A",
                      "title": "Left",
                      "width": 80,
                    }
                  }
                  overflow="clip"
                >
                  <HeaderTitle
                    className=""
                    col={0}
                    header={
                      Object {
                        "align": "left",
                        "id": "A",
                        "title": "Left",
                        "width": 80,
                      }
                    }
                    overflow="clip"
                  />
                </HeaderCell>,
                <HeaderCell
                  col={1}
                  header={
                    Object {
                      "align": "right",
                      "id": "B",
                      "title": "Right",
                    }
                  }
                  overflow="clip"
                >
                  <HeaderTitle
                    className=""
                    col={1}
                    header={
                      Object {
                        "align": "right",
                        "id": "B",
                        "title": "Right",
                      }
                    }
                    overflow="clip"
                  />
                </HeaderCell>,
                <HeaderCell
                  col={2}
                  header={
                    Object {
                      "id": "C",
                      "title": "Center",
                    }
                  }
                  overflow="clip"
                >
                  <HeaderTitle
                    className=""
                    col={2}
                    header={
                      Object {
                        "id": "C",
                        "title": "Center",
                      }
                    }
                    overflow="clip"
                  />
                </HeaderCell>,
                <HeaderCell
                  col={3}
                  header={
                    Object {
                      "id": "D",
                      "title": "Long Title that Might Overflow",
                    }
                  }
                  overflow="clip"
                >
                  <HeaderTitle
                    className=""
                    col={3}
                    header={
                      Object {
                        "id": "D",
                        "title": "Long Title that Might Overflow",
                      }
                    }
                    overflow="clip"
                  />
                </HeaderCell>,
              ],
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "$$actionCell",
              "nodeType": "host",
              "props": Object {
                "className": "cell read-only rdx-action-cell",
              },
              "ref": null,
              "rendered": null,
              "type": "th",
            },
            Object {
              "instance": null,
              "key": "header-A",
              "nodeType": "class",
              "props": Object {
                "children": <HeaderTitle
                  className=""
                  col={0}
                  header={
                    Object {
                      "align": "left",
                      "id": "A",
                      "title": "Left",
                      "width": 80,
                    }
                  }
                  overflow="clip"
                />,
                "col": 0,
                "header": Object {
                  "align": "left",
                  "id": "A",
                  "title": "Left",
                  "width": 80,
                },
                "overflow": "clip",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "A",
                "nodeType": "class",
                "props": Object {
                  "className": "",
                  "col": 0,
                  "header": Object {
                    "align": "left",
                    "id": "A",
                    "title": "Left",
                    "width": 80,
                  },
                  "overflow": "clip",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "header-B",
              "nodeType": "class",
              "props": Object {
                "children": <HeaderTitle
                  className=""
                  col={1}
                  header={
                    Object {
                      "align": "right",
                      "id": "B",
                      "title": "Right",
                    }
                  }
                  overflow="clip"
                />,
                "col": 1,
                "header": Object {
                  "align": "right",
                  "id": "B",
                  "title": "Right",
                },
                "overflow": "clip",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "B",
                "nodeType": "class",
                "props": Object {
                  "className": "",
                  "col": 1,
                  "header": Object {
                    "align": "right",
                    "id": "B",
                    "title": "Right",
                  },
                  "overflow": "clip",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "header-C",
              "nodeType": "class",
              "props": Object {
                "children": <HeaderTitle
                  className=""
                  col={2}
                  header={
                    Object {
                      "id": "C",
                      "title": "Center",
                    }
                  }
                  overflow="clip"
                />,
                "col": 2,
                "header": Object {
                  "id": "C",
                  "title": "Center",
                },
                "overflow": "clip",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "C",
                "nodeType": "class",
                "props": Object {
                  "className": "",
                  "col": 2,
                  "header": Object {
                    "id": "C",
                    "title": "Center",
                  },
                  "overflow": "clip",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "header-D",
              "nodeType": "class",
              "props": Object {
                "children": <HeaderTitle
                  className=""
                  col={3}
                  header={
                    Object {
                      "id": "D",
                      "title": "Long Title that Might Overflow",
                    }
                  }
                  overflow="clip"
                />,
                "col": 3,
                "header": Object {
                  "id": "D",
                  "title": "Long Title that Might Overflow",
                },
                "overflow": "clip",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "D",
                "nodeType": "class",
                "props": Object {
                  "className": "",
                  "col": 3,
                  "header": Object {
                    "id": "D",
                    "title": "Long Title that Might Overflow",
                  },
                  "overflow": "clip",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": "tr",
        },
        "type": "thead",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <tr>
            <td
              colSpan={4}
            >
              shallow
            </td>
          </tr>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <td
              colSpan={4}
            >
              shallow
            </td>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "shallow",
              "colSpan": 4,
            },
            "ref": null,
            "rendered": "shallow",
            "type": "td",
          },
          "type": "tr",
        },
        "type": "tbody",
      },
    ],
    "type": "table",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <thead>
            <tr>
              <th
                className="cell read-only rdx-action-cell"
              />
              <HeaderCell
                col={0}
                header={
                  Object {
                    "align": "left",
                    "id": "A",
                    "title": "Left",
                    "width": 80,
                  }
                }
                overflow="clip"
              >
                <HeaderTitle
                  className=""
                  col={0}
                  header={
                    Object {
                      "align": "left",
                      "id": "A",
                      "title": "Left",
                      "width": 80,
                    }
                  }
                  overflow="clip"
                />
              </HeaderCell>
              <HeaderCell
                col={1}
                header={
                  Object {
                    "align": "right",
                    "id": "B",
                    "title": "Right",
                  }
                }
                overflow="clip"
              >
                <HeaderTitle
                  className=""
                  col={1}
                  header={
                    Object {
                      "align": "right",
                      "id": "B",
                      "title": "Right",
                    }
                  }
                  overflow="clip"
                />
              </HeaderCell>
              <HeaderCell
                col={2}
                header={
                  Object {
                    "id": "C",
                    "title": "Center",
                  }
                }
                overflow="clip"
              >
                <HeaderTitle
                  className=""
                  col={2}
                  header={
                    Object {
                      "id": "C",
                      "title": "Center",
                    }
                  }
                  overflow="clip"
                />
              </HeaderCell>
              <HeaderCell
                col={3}
                header={
                  Object {
                    "id": "D",
                    "title": "Long Title that Might Overflow",
                  }
                }
                overflow="clip"
              >
                <HeaderTitle
                  className=""
                  col={3}
                  header={
                    Object {
                      "id": "D",
                      "title": "Long Title that Might Overflow",
                    }
                  }
                  overflow="clip"
                />
              </HeaderCell>
            </tr>
          </thead>,
          <tbody>
            <tr>
              <td
                colSpan={4}
              >
                shallow
              </td>
            </tr>
          </tbody>,
        ],
        "className": "foobar",
        "style": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <th
                className="cell read-only rdx-action-cell"
              />
              <HeaderCell
                col={0}
                header={
                  Object {
                    "align": "left",
                    "id": "A",
                    "title": "Left",
                    "width": 80,
                  }
                }
                overflow="clip"
              >
                <HeaderTitle
                  className=""
                  col={0}
                  header={
                    Object {
                      "align": "left",
                      "id": "A",
                      "title": "Left",
                      "width": 80,
                    }
                  }
                  overflow="clip"
                />
              </HeaderCell>
              <HeaderCell
                col={1}
                header={
                  Object {
                    "align": "right",
                    "id": "B",
                    "title": "Right",
                  }
                }
                overflow="clip"
              >
                <HeaderTitle
                  className=""
                  col={1}
                  header={
                    Object {
                      "align": "right",
                      "id": "B",
                      "title": "Right",
                    }
                  }
                  overflow="clip"
                />
              </HeaderCell>
              <HeaderCell
                col={2}
                header={
                  Object {
                    "id": "C",
                    "title": "Center",
                  }
                }
                overflow="clip"
              >
                <HeaderTitle
                  className=""
                  col={2}
                  header={
                    Object {
                      "id": "C",
                      "title": "Center",
                    }
                  }
                  overflow="clip"
                />
              </HeaderCell>
              <HeaderCell
                col={3}
                header={
                  Object {
                    "id": "D",
                    "title": "Long Title that Might Overflow",
                  }
                }
                overflow="clip"
              >
                <HeaderTitle
                  className=""
                  col={3}
                  header={
                    Object {
                      "id": "D",
                      "title": "Long Title that Might Overflow",
                    }
                  }
                  overflow="clip"
                />
              </HeaderCell>
            </tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": "$headers",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <th
                  className="cell read-only rdx-action-cell"
                />,
                Array [
                  <HeaderCell
                    col={0}
                    header={
                      Object {
                        "align": "left",
                        "id": "A",
                        "title": "Left",
                        "width": 80,
                      }
                    }
                    overflow="clip"
                  >
                    <HeaderTitle
                      className=""
                      col={0}
                      header={
                        Object {
                          "align": "left",
                          "id": "A",
                          "title": "Left",
                          "width": 80,
                        }
                      }
                      overflow="clip"
                    />
                  </HeaderCell>,
                  <HeaderCell
                    col={1}
                    header={
                      Object {
                        "align": "right",
                        "id": "B",
                        "title": "Right",
                      }
                    }
                    overflow="clip"
                  >
                    <HeaderTitle
                      className=""
                      col={1}
                      header={
                        Object {
                          "align": "right",
                          "id": "B",
                          "title": "Right",
                        }
                      }
                      overflow="clip"
                    />
                  </HeaderCell>,
                  <HeaderCell
                    col={2}
                    header={
                      Object {
                        "id": "C",
                        "title": "Center",
                      }
                    }
                    overflow="clip"
                  >
                    <HeaderTitle
                      className=""
                      col={2}
                      header={
                        Object {
                          "id": "C",
                          "title": "Center",
                        }
                      }
                      overflow="clip"
                    />
                  </HeaderCell>,
                  <HeaderCell
                    col={3}
                    header={
                      Object {
                        "id": "D",
                        "title": "Long Title that Might Overflow",
                      }
                    }
                    overflow="clip"
                  >
                    <HeaderTitle
                      className=""
                      col={3}
                      header={
                        Object {
                          "id": "D",
                          "title": "Long Title that Might Overflow",
                        }
                      }
                      overflow="clip"
                    />
                  </HeaderCell>,
                ],
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "$$actionCell",
                "nodeType": "host",
                "props": Object {
                  "className": "cell read-only rdx-action-cell",
                },
                "ref": null,
                "rendered": null,
                "type": "th",
              },
              Object {
                "instance": null,
                "key": "header-A",
                "nodeType": "class",
                "props": Object {
                  "children": <HeaderTitle
                    className=""
                    col={0}
                    header={
                      Object {
                        "align": "left",
                        "id": "A",
                        "title": "Left",
                        "width": 80,
                      }
                    }
                    overflow="clip"
                  />,
                  "col": 0,
                  "header": Object {
                    "align": "left",
                    "id": "A",
                    "title": "Left",
                    "width": 80,
                  },
                  "overflow": "clip",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": "A",
                  "nodeType": "class",
                  "props": Object {
                    "className": "",
                    "col": 0,
                    "header": Object {
                      "align": "left",
                      "id": "A",
                      "title": "Left",
                      "width": 80,
                    },
                    "overflow": "clip",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "header-B",
                "nodeType": "class",
                "props": Object {
                  "children": <HeaderTitle
                    className=""
                    col={1}
                    header={
                      Object {
                        "align": "right",
                        "id": "B",
                        "title": "Right",
                      }
                    }
                    overflow="clip"
                  />,
                  "col": 1,
                  "header": Object {
                    "align": "right",
                    "id": "B",
                    "title": "Right",
                  },
                  "overflow": "clip",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": "B",
                  "nodeType": "class",
                  "props": Object {
                    "className": "",
                    "col": 1,
                    "header": Object {
                      "align": "right",
                      "id": "B",
                      "title": "Right",
                    },
                    "overflow": "clip",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "header-C",
                "nodeType": "class",
                "props": Object {
                  "children": <HeaderTitle
                    className=""
                    col={2}
                    header={
                      Object {
                        "id": "C",
                        "title": "Center",
                      }
                    }
                    overflow="clip"
                  />,
                  "col": 2,
                  "header": Object {
                    "id": "C",
                    "title": "Center",
                  },
                  "overflow": "clip",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": "C",
                  "nodeType": "class",
                  "props": Object {
                    "className": "",
                    "col": 2,
                    "header": Object {
                      "id": "C",
                      "title": "Center",
                    },
                    "overflow": "clip",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "header-D",
                "nodeType": "class",
                "props": Object {
                  "children": <HeaderTitle
                    className=""
                    col={3}
                    header={
                      Object {
                        "id": "D",
                        "title": "Long Title that Might Overflow",
                      }
                    }
                    overflow="clip"
                  />,
                  "col": 3,
                  "header": Object {
                    "id": "D",
                    "title": "Long Title that Might Overflow",
                  },
                  "overflow": "clip",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": "D",
                  "nodeType": "class",
                  "props": Object {
                    "className": "",
                    "col": 3,
                    "header": Object {
                      "id": "D",
                      "title": "Long Title that Might Overflow",
                    },
                    "overflow": "clip",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
            ],
            "type": "tr",
          },
          "type": "thead",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <td
                colSpan={4}
              >
                shallow
              </td>
            </tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <td
                colSpan={4}
              >
                shallow
              </td>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "shallow",
                "colSpan": 4,
              },
              "ref": null,
              "rendered": "shallow",
              "type": "td",
            },
            "type": "tr",
          },
          "type": "tbody",
        },
      ],
      "type": "table",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
